compiler/rustc_passes/src/dead.rs

```diff
@@ -857,6 +857,31 @@ fn check_item<'tcx>(
             // global_asm! is always live.
             worklist.push((id.owner_id.def_id, ComesFromAllowExpect::No));
         }
+        DefKind::OpaqueTy => {
+            if let hir::ItemKind::OpaqueTy(ty) = tcx.hir().item(id).kind {
+                for bound in ty.bounds {
+                    if let hir::GenericBound::Trait(polytraitref, _) = bound
+                        && let Some(pathsegment) = polytraitref.trait_ref.path.segments.last()
+                        && let Some(args) = pathsegment.args
+                    {
+                        for constraint in args.constraints {
+                            if let Some(item) = tcx
+                                .associated_items(pathsegment.res.def_id())
+                                .filter_by_name_unhygienic(constraint.ident.name)
+                                .find(|i| {
+                                    i.kind == ty::AssocKind::Const
+                                        && i.ident(tcx).normalize_to_macros_2_0()
+                                            == constraint.ident
+                                })
+                                && let Some(local_def_id) = item.def_id.as_local()
+                            {
+                                worklist.push((local_def_id, ComesFromAllowExpect::No));
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
I don’t think so, because the `def_id` of the associated item actually doesn’t show up anywhere in the hir of the ty. I could totally be missing something, though.

Reply:
Yea, this is a more fundamental problem. https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=dd0c2930c397acba2692b51a633b367d will still fail after your PR. If these could be used in trait objects we'd probably see the issue there, too.

Considering that

```rust
#![deny(dead_code)]

trait Tr {
    type T;
}

impl Tr for () {
    type T = ();
}

fn foo<T: Tr>(_: T) { } 


fn main() {
    foo(());
}
```

passes, I'm assuming that if a trait is marked as used, all its assoc types are marked as used. We should do the same for assoc consts, not sure how that didn't happen or if there was an explicit reason for it.

Reply:
Looks like linting for associated consts was added recently in #125572, cc @mu001999 @pnkfelix

You're right that if a trait is marked as used, all its assoc types are marked as used.

Reply:
> I'm assuming that if a trait is marked as used, all its assoc types are marked as used.

It was made in #126618. I think it's not the best solution, i.e., marking assoc const/tys live if the trait is used, considering that
```rust
trait Trait {
    const X: i32;
    fn foo(&self) {}
}

impl Trait for () {
    const X: i32 = 0;
}

fn main() {
    ().foo();
}
```
The assoc const is not used and we can lint these now.